llama/llama.go

```diff
@@ -680,3 +682,35 @@ func (s *SamplingContext) Sample(llamaContext *Context, idx int) int {
 func (s *SamplingContext) Accept(id int, applyGrammar bool) {
 	C.gpt_sampler_caccept(s.c, C.llama_token(id), C.bool(applyGrammar))
 }
+
+type JsonSchema struct {
+	Defs       map[string]any `json:"$defs,omitempty"`
+	Properties map[string]any `json:"properties,omitempty"`
+	Required   []string       `json:"required,omitempty"`
+	Title      string         `json:"title,omitempty"`
+	Type       string         `json:"type,omitempty"`
+}
+
+func (js JsonSchema) AsGrammar() string {
+	// Convert the JSON schema to a string representation
+	jsonBytes, err := json.Marshal(js)
+	if err != nil {
+		return ""
+	}
+
+	cStr := C.CString(string(jsonBytes))
+	defer C.free(unsafe.Pointer(cStr))
+
+	// Allocate buffer for grammar output with reasonable size
+	const maxLen = 32768 // 32KB
+	buf := make([]byte, maxLen)
+
+	// Call C function to convert schema to grammar
+	length := C.schema_to_grammar(cStr, (*C.char)(unsafe.Pointer(&buf[0])), C.int32_t(maxLen))
+	if length <= 0 {
+		slog.Warn("unable to convert schema to grammar")
+		return ""
+	}
+
+	return string(C.GoBytes(unsafe.Pointer(&buf[0]), length))
```

Comment:
This unnecessary since `buf` is `[]byte`

```suggestion
	return string(buf[:length])
```

Reply:
oh what the heck i thought i had changed that - updating